module/zfs/dmu_send.c

```diff
@@ -1537,7 +1541,8 @@ restore_object(struct restorearg *ra, objset_t *os, struct drr_object *drro)
 		}
 	}
 
-	tx = dmu_tx_create(os);
+	count = DN_BONUSLEN_TO_COUNT(drro->drr_bonuslen);
+	tx = dmu_tx_create_dn_count(os, count);
```

Comment:
@ahrens my understanding is that what your suggesting is what implemented in the initial version of the patch.  However, that ended up being problematic because it's possible that the preferred dnode size could change mid-way through a transaction.  Adding it to the tx seemed like the cleanest way to ensure the same dnode size was used all the way through the tx.  This does seem like it should be done in dmu_tx_hold_*().